flink-runtime/src/main/java/org/apache/flink/runtime/state/KeyedStateFactory.java

```diff
@@ -66,4 +66,32 @@ <N, SV, SEV, S extends State, IS extends S> IS createInternalState(
             @Nonnull StateDescriptor<S, SV> stateDesc,
             @Nonnull StateSnapshotTransformFactory<SEV> snapshotTransformFactory)
             throws Exception;
+
+    /**
+     * Creates and returns a new {@link InternalKvState}.
+     *
+     * @param namespaceSerializer TypeSerializer for the state namespace.
+     * @param stateDesc The {@code StateDescriptor} that contains the name of the state.
+     * @param snapshotTransformFactory factory of state snapshot transformer.
+     * @param allowFutureMetadataUpdates whether allow metadata to update in the future or not.
+     * @param <N> The type of the namespace.
+     * @param <SV> The type of the stored state value.
+     * @param <SEV> The type of the stored state value or entry for collection types (list or map).
+     * @param <S> The type of the public API state.
+     * @param <IS> The type of internal state.
+     */
+    @Nonnull
+    default <N, SV, SEV, S extends State, IS extends S> IS createInternalState(
```

Comment:
This method has createOrUpdate semantics, right?
(as well as the methods down the call chain)
If so, do you think they should be named accordingly?

Reply:
I think so, and I think it deserves a seperate commit about refactoring. WDYT?

Reply:
I agree, separate commit would be more clear.